frontend/src/main/scala/bloop/io/DirChecksum.scala

```diff
@@ -0,0 +1,92 @@
+package bloop.io
+
+import java.io.{InputStream, SequenceInputStream}
+import java.nio.file.Files
+import java.nio.file.attribute.FileTime
+import java.util.Enumeration
+import java.util.zip.{Adler32, CheckedInputStream}
+
+/**
+ * Represents the state of a directory and some files in which we track changes.
+ *
+ * @param dir           The directory to track.
+ * @param pattern       The pattern matching the files that must be tracked inside `dir`.
+ * @param modifiedTimes The last modification time of the tracked files.
+ * @param contentsChecksum The checksum of all the contents of the directory.
+ */
+final class DirChecksum(dir: AbsolutePath,
+                        pattern: String,
+                        private[this] var modifiedTimes: List[(AbsolutePath, FileTime)],
+                        contentsChecksum: Long) {
+
+  /**
+   * Inspects the directory for changes.
+   *
+   * @return true if the directory has changed, false otherwise.
+   */
+  def changed(): Boolean = {
+    val newModifiedTimes = DirChecksum.getFiles(dir, pattern)
+    if (newModifiedTimes == modifiedTimes) false
+    else
+      synchronized {
```

Comment:
Is synchronizing here working? We're only syncing on the instance of `DirChecksum`.